core/lib/makeswift/page.tsx

```diff
@@ -0,0 +1,35 @@
+import { Page as MakeswiftPage } from '@makeswift/runtime/next';
+import { notFound } from 'next/navigation';
+
+import { locales } from '~/i18n/routing';
+
+import { client, getPageSnapshot } from './client';
+
+interface PageParams {
+  locale: string;
+  rest: string[];
+}
+
+function localesFanOut(path: string): PageParams[] {
+  return locales.map((locale) => ({
+    rest: path.split('/').filter((segment) => segment !== ''),
+    locale,
+  }));
+}
+
+export async function generateStaticParams(): Promise<PageParams[]> {
+  const pages = await client.getPages().toArray();
+
+  return pages.filter((page) => page.path !== '/').flatMap((page) => localesFanOut(page.path));
+}
```

Comment:
This feels like it should be kept in the page file since it's coupled to the route and where the page file is in the filesystem.

Reply:
> This feels like it should be kept in the page file since it's coupled to the route

Yeah, I debated it, but I think the implementation is going to be the same sans the page prefix (which I forgot to pass here).

> where the page file is in the filesystem

It's re-exported from the page file.

Reply:
On second thought, I think I see what you mean: specifically the `rest` param is page-location specific.